pkg/api/query/v1.go

```diff
@@ -1037,6 +1049,48 @@ func (qapi *QueryAPI) stores(_ *http.Request) (interface{}, []error, *api.ApiErr
 	return statuses, nil, nil, func() {}
 }
 
+func (qapi *QueryAPI) getLabelMatchers(matchers []string, tenant string) ([][]*labels.Matcher, *api.ApiError) {
+	tenantLabelMatcher := &labels.Matcher{
+		Name:  qapi.tenantLabel,
+		Type:  labels.MatchEqual,
+		Value: tenant,
+	}
+
+	var matcherSets [][]*labels.Matcher
```

Comment:
We can preallocate using make([][]*labels.Matcher, len(matchers))

Reply:
Sometimes we also have `len(matchers)+1` though depending on the situation. Tried out a few things, but didn't manage to make it work well, not exactly sure what went wrong though. Can investigate further, if you think preallocation is a big improvement.

Reply:
It is by no mean a performance critical operation here as the number of matchers should be fairly low on most operations. But doesn't cost anything to do `matcherSets := make([][]*labels.Matcher, 0, len(matchers))`. I am surprised it does not work. But it is not a blocker, you can leave it as is if you prefer.